README.md
# Release B Demo

## Warning

This repository is not designed for productive usage. Instead it should be used for demonstration and learning purposes.

## Note
Release B will not be published in its entirety at once. Currently the following Modules are available:
* Asset Registry
  * [origin-asset-registry-contracts](https://github.com/energywebfoundation/origin-asset-registry-contracts)
  * [origin-asset-registry-lib](https://github.com/energywebfoundation/origin-asset-registry-lib)
* User Registry
  * [origin-user-registry-contracts](https://github.com/energywebfoundation/origin-user-registry-contracts)
* Utils
  * [origin-utils-general-precise-proofs](https://github.com/energywebfoundation/origin-utils-general-precise-proofs) 
  * [origin-utils-general-contracts](https://github.com/energywebfoundation/origin-utils-general-contracts)
  * [origin-utils-demo](https://github.com/energywebfoundation/origin-utils-demo) 
  * [origin-utils-general-lib](https://github.com/energywebfoundation/origin-utils-general-lib) 
  * [origin-utils-testbackend](https://github.com/energywebfoundation/origin-utils-testbackend) 

Coming soon:
* Origin Issuer
* Origin Market

## Getting Started

1. Install dependencies `npm install`
2. Start ganache `npm run start-ganache`
3. Start test backend `npm run start-test-backend`
4. Build demo code `npm run build-ts`
5. Start demo `npm start`
6. Use [origin-asset-registry-frontend](https://github.com/energywebfoundation/origin-asset-registry-frontend) to see the assets.



changelog.md
# origin release B

## general
* split up the repositories
    * ew-asset-registry-contracts
    * ew-asset-registry-lib
    * ew-asset-registry-frontend
    * ew-market-contracts
    * ew-market-lib
    * ew-origin-contracts
    * ew-origin-lib
    * ew-user-registry-contracts
    * ew-user-lib
    * ew-utils-general-contracts
    * ew-utils-general-lib  
    * ew-utils-general-precise-proofs
    * ew-utils-demo
    * ew-utils-deployment 
    * ew-utils-testbackend
    * ew-utils-testtoken 

* moved many onchain-properties to be stored offchain 
* created simple backend for testing origin
* updated all contracts to Solidity > 0.5.0
* smart contracts are backwards compatible to truffle 

## asset-registry
* created separate repositories for assets: 
    * ew-asset-registry-contracts
    * ew-asset-registry-lib
    * ew-asset-registry-frontend
### smart contracts 
* created "AssetContractLookup" Smart Contract: smart contract that stores the address of the actual logic contracts (can be used for ENS)
* refactored code, using structs for transferring information between logic and db contract
* moved onchain-properties to be stored offchain: 
    * operationalSince
    * capacityWh
    * country
    * region
    * zip 
    * street
    * houseNumber
    * gpsLatitude
    * gpsLongitude
* assets can be linked to one OriginContract, allowing him to createTradableEntities (e.g. certificates)
* without such a link only the produced energy gets tracked, no certificates will be created (only energy produced will be tracked)
* enforced rule that each smartMeter-address can be used once
* assets can be retrieved by its asstetId or by its smartMeter-address (both are working)
* each asset supports up to 10 matcher 
* created interfaces to talk to the AssetProducingRegistryLogic andAssetConsumingRegistryLogic from outside (e.g. through origin smart contracts)
* created npm-package for ew-asset-registry-contracts
    * typescript bindings for smart contracts
    * function to deploy asset-registry-contracts
    
### asset-lib    
* precise proofs are used for storing and retrieving the offchain properties
* CO2 saved is not stored anymore, gets calculated instead
* changed library functions to both support a privateKey or an address when sending a transaction
* created npm-package for ew-asset-registry-lib 

## market-registry
* created separate repositories for market: 
    * ew-market-contracts
    * ew-market-lib
### smart contracts
* created "MarketContractLookup" contract: smart contract that stores the address of the current logic contract (can be used for ENS)
* matcher is allowed to change MatcherOffChainProperties
* separated properties in on- and offchain
* changed the way agreements are working
* both supply and demand have to exists
* supply:
    * producing asset
    * only owner of assets can create supplies
    * moved details of supply to offchain
* demand:
    * amount of energy someone wants to have in a certain amount of time
    * moved details of demands to offchain
* agreement: 
    * agreement = existing supply and demand
    * both the owner of a supply (= asset-owner) and owner of a demand (= trader) have to approve an agreement before it gets activated
* created npm-package for ew-market-registry
    * typescript bindings for smart contracts
    * function to deploy market contracts

### market lib
* precise proofs are used to storing and receiving offchain properties for demand, supply and agreement
* changed library functions to both support a privateKey or an address when sending a transaction

* new offchain properties for agreements
    * start
    * end
    * price
    * currency 
    * period
    * timeFrame

* new offchain properties for matcherOffchain
    * currentWh
    * currentPeriod

* new offchain properties for demand
    * timeFrame
    * pricePerCertificateWh
    * currency
    * producingAsset (optional)
    * consumingAsset (optional)
    * locationCountry (optional)
    * locationRegion (optional)
    * assetType (optional)
    * minCO2Offset (optional)
    * otherGreenAttributes (optional)
    * typeOfPublicSupport (optional)
    * targetWhPerPeriod
    * registryCompliance (optional)
* created npm-package for ew-market-lib

## origin-registry
* created separate repositories for origin (certificate / bundle):
    * ew-origin-contracts
    * ew-origin-lib
### smart contracts
* created "OriginContractLookup" contract: smart contract that stores the address of the current logic contract (can be used for ENS)
* logging energy will create a TradableEntity
    * owner of the producing asset will be the first owner of the created entity
    * TradableEntities support up to 10 matcher
    * TradableEntity are ERC721 and ERC165 compliant
    * added ERC20-contract address and directPurchasePrice to a tradableEntity -> allows direct buying of a TradableEntity (when ERC20 token transfer is finished)
* TradableEntities can either be stand-alone TradableEntities, Certificates or CertificateEnergyBundle
* Certificates:
    * assetOwner can set a maximum amount of owner changes for certificates
    * auto-retirement of certificates when the maximum amount of owner changes is reached
    * splitting of certificates: 
        * a certificate can be split into to child certificates which will inherit its parent properties (and could be split again)
        * once a certificate is split it cannot be traded / split / retired anymore
* created npm-package for ew-market-registry
    * typescript bindings for smart contracts
    * function to deploy contracts (as certificate or bundle of certificates and energy)

### origin lib
* changed library functions to both support a privateKey or an address when sending a transaction

## user-registry
* created separate repositories for the user-registry
    * ew-user-contracts
    * ew-user-lib
### smart contracts

src/demo.ts
import * as fs from 'fs';
import Web3 = require('web3');
import { certificateDemo } from './certificate'
import { deployEmptyContracts } from './deployEmpty'
import { logger } from './Logger';

import * as Certificate from 'ew-origin-lib';
import * as User from 'ew-user-registry-lib'
import * as Asset from 'ew-asset-registry-lib'
import * as GeneralLib from 'ew-utils-general-lib';
import * as Market from 'ew-market-lib';

import { UserContractLookup, UserLogic } from 'ew-user-registry-contracts';
import { AssetContractLookup, AssetProducingRegistryLogic, AssetConsumingRegistryLogic } from 'ew-asset-registry-contracts';
import { OriginContractLookup, CertificateLogic, migrateCertificateRegistryContracts } from 'ew-origin-contracts';
import { deployERC20TestToken, Erc20TestToken, TestReceiver, deployERC721TestReceiver } from 'ew-erc-test-contracts';
import { MarketContractLookup, MarketLogic } from 'ew-market-contracts';


function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}



export const marketDemo = async() => {

  const startTime = Date.now()

  await deployEmptyContracts()

  const connectionConfig = JSON.parse(fs.readFileSync(process.cwd() + '/connection-config.json', 'utf8').toString());
  const demoConfig = JSON.parse(fs.readFileSync(process.cwd() + '/demo-config.json', 'utf8').toString());
  const contractConfig = JSON.parse(fs.readFileSync(process.cwd() + '/contractConfig.json', 'utf8').toString());

  const web3 = new Web3(connectionConfig.develop.web3);

  const adminPK = demoConfig.topAdminPrivateKey.startsWith('0x') ?
      demoConfig.topAdminPrivateKey : '0x' + demoConfig.topAdminPrivateKey;

  const matcherPK = demoConfig.matcherPrivateKey.startsWith('0x') ?
      demoConfig.matcherPrivateKey : '0x' + demoConfig.matcherPrivateKey;

  const adminAccount = web3.eth.accounts.privateKeyToAccount(adminPK);
  const matcherAccount = web3.eth.accounts.privateKeyToAccount(matcherPK)

  //create logic instances
  const userLogic = new UserLogic(web3, contractConfig.userLogic)
  const assetProducingRegistryLogic = new AssetProducingRegistryLogic(web3, contractConfig.assetProducingRegistryLogic)
  const assetConsumingRegistryLogic = new AssetConsumingRegistryLogic(web3, contractConfig.assetConsumingRegistryLogic)
  const certificateLogic = new CertificateLogic(web3, contractConfig.certificateLogic)
  const marketLogic = new MarketLogic(web3, contractConfig.marketLogic)

  //set the admin account as an asset admin
  await userLogic.setUser(adminAccount.address, 'admin', { privateKey: adminPK });
  await userLogic.setRoles(adminAccount.address, 3, { privateKey: adminPK });

  //initialize a token to handle demo erc20 trading
  let erc20TestToken: Erc20TestToken

  //initialize variables for storing timeframe and currency
  let timeFrame
  let currency

  //blockchain configuration
  let conf: GeneralLib.Configuration.Entity;

  conf = {
      blockchainProperties: {
          activeUser: {
              address: adminAccount.address, privateKey: adminPK,
          },
          producingAssetLogicInstance: assetProducingRegistryLogic,
          consumingAssetLogicInstance: assetConsumingRegistryLogic,
          certificateLogicInstance: certificateLogic,
          userLogicInstance: userLogic,
          marketLogicInstance: marketLogic,
          web3,
      },
      offChainDataSource: {
          baseUrl: 'http://localhost:3030',
      },
      logger,
  };

  const actionsArray = demoConfig.flow

  for(const action of actionsArray) {
    switch(action.type) {
      case "CREATE_DEMAND":

        console.log("-----------------------------------------------------------")

        conf.blockchainProperties.activeUser = {
          address: action.data.trader, privateKey: action.data.traderPK,
        };

        let assetTypeConfig

        switch (action.data.assettype) {
            case 'Wind': assetTypeConfig = GeneralLib.AssetType.Wind
                break
            case 'Solar': assetTypeConfig = GeneralLib.AssetType.Solar
                break
            case 'RunRiverHydro': assetTypeConfig = GeneralLib.AssetType.RunRiverHydro
                break
            case 'BiomassGas': assetTypeConfig = GeneralLib.AssetType.BiomassGas
        }

        let assetCompliance

        switch (action.data.registryCompliance) {
            case 'IREC': assetCompliance = GeneralLib.Compliance.IREC
                break
            case 'EEC': assetCompliance = GeneralLib.Compliance.EEC
                break
            case 'TIGR': assetCompliance = GeneralLib.Compliance.TIGR
                break
            default:
                assetCompliance = GeneralLib.Compliance.none
                break
        }

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly
            break
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly
            break
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily
            break
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly
            break
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro
            break
          case 'USD':
            currency = GeneralLib.Currency.USD
            break
          case 'Ether':
            currency = GeneralLib.Currency.Ether
            break
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar
            break
        }

        const demandOffchainProps: Market.Demand.DemandOffchainproperties = {
            timeframe: timeFrame,
            pricePerCertifiedWh: action.data.pricePerCertifiedWh,
            currency: currency,
            productingAsset: action.data.producingAsset,
            consumingAsset: action.data.consumingAsset,
            locationCountry: action.data.country,
            locationRegion: action.data.region,
            assettype: assetTypeConfig,
            minCO2Offset: action.data.minCO2Offset,
            otherGreenAttributes: action.data.otherGreenAttributes,
            typeOfPublicSupport: action.data.typeOfPublicSupport,
            targetWhPerPeriod: action.data.targetWhPerPeriod,
            registryCompliance: assetCompliance,
        };

        const demandProps: Market.Demand.DemandOnChainProperties = {
          url: null,
          propertiesDocumentHash: null,
          demandOwner: action.data.trader,
        };

        try {
          const demand = await Market.Demand.createDemand(demandProps, demandOffchainProps, conf);
          delete demand.proofs;
          delete demand.configuration;
          console.log("Demand Created, ID: " + demand.id)
        } catch(e) {
          console.log("Error creating a demand\n" + e)
        }

        console.log("-----------------------------------------------------------\n")

        break
      case "CREATE_SUPPLY":
        console.log("-----------------------------------------------------------")

        conf.blockchainProperties.activeUser = {
          address: action.data.assetOwner, privateKey: action.data.assetOwnerPK,
        };

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly
            break
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly
            break
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily
            break
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly
            break
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro
            break
          case 'USD':
            currency = GeneralLib.Currency.USD
            break
          case 'Ether':
            currency = GeneralLib.Currency.Ether
            break
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar
            break
        }

        const supplyOffChainProperties: Market.Supply.SupplyOffchainProperties = {
            price: action.data.price,
            currency: currency,
            availableWh: action.data.availableWh,
            timeframe: timeFrame,
        };

        const supplyProps: Market.Supply.SupplyOnChainProperties = {
          url: null,
          propertiesDocumentHash: null,
          assetId: action.data.assetId,
        };


        try {
          const supply = await Market.Supply.createSupply(supplyProps, supplyOffChainProperties, conf);
          delete supply.proofs;
          delete supply.configuration;
          console.log("Onboarded Supply ID: " + supply.id)
        } catch(e) {
          console.log("Error onboarding supply\n" + e)
        }

        console.log("-----------------------------------------------------------\n")

        break

      case "MAKE_AGREEMENT":
        console.log("-----------------------------------------------------------")

        conf.blockchainProperties.activeUser = {
          address: action.data.creator,
          privateKey: action.data.creatorPK,
        };

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly
            break
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly
            break
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily
            break
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly
            break
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro
            break
          case 'USD':
            currency = GeneralLib.Currency.USD
            break
          case 'Ether':
            currency = GeneralLib.Currency.Ether
            break
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar
            break
        }

        const agreementOffchainProps: Market.Agreement.AgreementOffChainProperties = {
          start: action.data.startTime,
          ende: action.data.endTime,
          price: action.data.price,
          currency: currency,
          period: action.data.period,
          timeframe: timeFrame,
        };

        const matcherOffchainProps: Market.Agreement.MatcherOffchainProperties = {
          currentWh: action.data.currentWh,
          currentPeriod: action.data.currentPeriod,
        };

        const agreementProps: Market.Agreement.AgreementOnChainProperties = {
          propertiesDocumentHash: null,
          url: null,
          matcherDBURL: null,
          matcherPropertiesDocumentHash: null,
          demandId: action.data.demandId,
          supplyId: action.data.supplyId,
          allowedMatcher: [action.data.allowedMatcher],
        };

        try {
          const agreement = await Market.Agreement.createAgreement(agreementProps, agreementOffchainProps, matcherOffchainProps, conf);
          delete agreement.proofs;
          delete agreement.configuration;
          delete agreement.propertiesDocumentHash;
          delete agreement.matcherPropertiesDocumentHash;
          if(agreement.approvedBySupplyOwner && agreement.approvedByDemandOwner){
            console.log("Agreement Confirmed")
          }
          else if(!agreement.approvedByDemandOwner) {
            console.log("Demand Owner did not approve yet")
          }
          else if(!agreement.approvedBySupplyOwner) {
            console.log("Supply Owner did not approve yet")
          }
        } catch(e) {
          console.log("Error making an agreement\n" + e)
        }

        console.log("-----------------------------------------------------------\n")
        break
      case "APPROVE_AGREEMENT":
        console.log("-----------------------------------------------------------")

        conf.blockchainProperties.activeUser = {
          address: action.data.agree,
          privateKey: action.data.agreePK,
        };

        try {
          let agreement: Market.Agreement.Entity = await (new Market.Agreement.Entity('0', conf)).sync();
          await agreement.approveAgreementSupply();
          agreement = await agreement.sync();
          if(agreement.approvedBySupplyOwner && agreement.approvedByDemandOwner){
            console.log("Agreement Confirmed")
          }
          else if(!agreement.approvedByDemandOwner) {
            console.log("Demand Owner did not approve yet")
          }
          else if(!agreement.approvedBySupplyOwner) {
            console.log("Supply Owner did not approve yet")
          }
        } catch(e) {
          console.log("Could not approve agreement\n" + e)
        }

        console.log("-----------------------------------------------------------\n")
        break
      default:
        const passString = JSON.stringify(action)
        await certificateDemo(passString, conf, adminPK)
    }
  }
  console.log("Total Time: " + ((Date.now()-startTime)/1000) + " seconds")
}

marketDemo()


src/market.ts
// Copyright 2018 Energy Web Foundation
// This file is part of the Origin Application brought to you by the Energy Web Foundation,
// a global non-profit organization focused on accelerating blockchain technology across the energy sector,
// incorporated in Zug, Switzerland.
//
// The Origin Application is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// This is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY and without an implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details, at <http://www.gnu.org/licenses/>.
//
// @authors: slock.it GmbH; Heiko Burkhardt, heiko.burkhardt@slock.it; Martin Kuechler, martin.kuchler@slock.it; Chirag Parmar, chirag.parmar@slock.it

import * as fs from 'fs';
import Web3 from 'web3';
import { certificateDemo } from './certificate'
import { deployEmptyContracts } from './deployEmpty'
import { logger } from './Logger';

import * as Certificate from 'ew-origin-lib';
import * as User from 'ew-user-registry-lib'
import * as Asset from 'ew-asset-registry-lib'
import * as GeneralLib from 'ew-utils-general-lib';
import * as Market from 'ew-market-lib';

import { UserContractLookup, UserLogic } from 'ew-user-registry-contracts';
import { AssetContractLookup, AssetProducingRegistryLogic, AssetConsumingRegistryLogic } from 'ew-asset-registry-contracts';
import { OriginContractLookup, CertificateLogic, migrateCertificateRegistryContracts } from 'ew-origin-contracts';
import { deployERC20TestToken, Erc20TestToken, TestReceiver, deployERC721TestReceiver } from 'ew-erc-test-contracts';
import { MarketContractLookup, MarketLogic } from 'ew-market-contracts';


function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}



export const marketDemo = async (demoFile?: string) => {

  const startTime = Date.now();

  const connectionConfig = JSON.parse(fs.readFileSync('./connection-config.json', 'utf8').toString());

  let demoConfig;
  if (!demoFile) demoConfig = JSON.parse(fs.readFileSync('./config/demo-config.json', 'utf8').toString());
  else demoConfig = JSON.parse(demoFile);

  const contractConfig = JSON.parse(fs.readFileSync('./config/contractConfig.json', 'utf8').toString());

  const web3 = new Web3(connectionConfig.develop.web3);

  const adminPK = demoConfig.topAdminPrivateKey.startsWith('0x') ?
    demoConfig.topAdminPrivateKey : '0x' + demoConfig.topAdminPrivateKey;

  const adminAccount = web3.eth.accounts.privateKeyToAccount(adminPK);

  //create logic instances
  const userLogic = new UserLogic(web3, contractConfig.userLogic)
  const assetProducingRegistryLogic = new AssetProducingRegistryLogic(web3, contractConfig.assetProducingRegistryLogic)
  const assetConsumingRegistryLogic = new AssetConsumingRegistryLogic(web3, contractConfig.assetConsumingRegistryLogic)
  const certificateLogic = new CertificateLogic(web3, contractConfig.certificateLogic)
  const marketLogic = new MarketLogic(web3, contractConfig.marketLogic)

  // set the admin account as an asset admin
  await userLogic.setUser(adminAccount.address, 'admin', { privateKey: adminPK });
  await userLogic.setRoles(adminAccount.address, 3, { privateKey: adminPK });

  // initialize a token to handle demo erc20 trading
  let erc20TestToken: Erc20TestToken

  // initialize variables for storing timeframe and currency
  let timeFrame
  let currency

  // blockchain configuration
  let conf: GeneralLib.Configuration.Entity;

  conf = {
    blockchainProperties: {
      activeUser: {
        address: adminAccount.address, privateKey: adminPK,
      },
      producingAssetLogicInstance: assetProducingRegistryLogic,
      consumingAssetLogicInstance: assetConsumingRegistryLogic,
      certificateLogicInstance: certificateLogic,
      userLogicInstance: userLogic,
      marketLogicInstance: marketLogic,
      web3,
    },
    offChainDataSource: {
      baseUrl: 'http://localhost:3030',
    },
    logger,
  };

  const actionsArray = demoConfig.flow

  for (const action of actionsArray) {
    switch (action.type) {
      case 'CREATE_DEMAND':

        console.log('-----------------------------------------------------------');

        conf.blockchainProperties.activeUser = {
          address: action.data.trader, privateKey: action.data.traderPK,
        };

        let assetTypeConfig;

        switch (action.data.assettype) {
          case 'Wind': assetTypeConfig = GeneralLib.AssetType.Wind;
                       break;
          case 'Solar': assetTypeConfig = GeneralLib.AssetType.Solar;
                        break;
          case 'RunRiverHydro': assetTypeConfig = GeneralLib.AssetType.RunRiverHydro;
                                break;
          case 'BiomassGas': assetTypeConfig = GeneralLib.AssetType.BiomassGas;
        }

        let assetCompliance;

        switch (action.data.registryCompliance) {
          case 'IREC': assetCompliance = GeneralLib.Compliance.IREC;
                       break;
          case 'EEC': assetCompliance = GeneralLib.Compliance.EEC;
                      break;
          case 'TIGR': assetCompliance = GeneralLib.Compliance.TIGR;
                       break;
          default:
            assetCompliance = GeneralLib.Compliance.none;
            break;
        }

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly;
            break;
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly;
            break;
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily;
            break;
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly;
            break;
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro;
            break;
          case 'USD':
            currency = GeneralLib.Currency.USD;
            break;
          case 'Ether':
            currency = GeneralLib.Currency.Ether;
            break;
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar;
            break;
        }

        const demandOffchainProps: Market.Demand.DemandOffchainproperties = {
          timeframe: timeFrame,
          pricePerCertifiedWh: action.data.pricePerCertifiedWh,
          currency,
          productingAsset: action.data.producingAsset,
          consumingAsset: action.data.consumingAsset,
          locationCountry: action.data.country,
          locationRegion: action.data.region,
          assettype: assetTypeConfig,
          minCO2Offset: action.data.minCO2Offset,
          otherGreenAttributes: action.data.otherGreenAttributes,
          typeOfPublicSupport: action.data.typeOfPublicSupport,
          targetWhPerPeriod: action.data.targetWhPerPeriod,
          registryCompliance: assetCompliance,
        };

        const demandProps: Market.Demand.DemandOnChainProperties = {
          url: '',
          propertiesDocumentHash: '',
          demandOwner: action.data.trader,
        };

        try {
          const demand = await Market.Demand.createDemand(demandProps, demandOffchainProps, conf);
          delete demand.proofs;
          delete demand.configuration;
          conf.logger.info('Demand Created, ID: ' + demand.id);
        } catch (e) {
          conf.logger.error('Demand could not be created\n' + e);
        }

        console.log('-----------------------------------------------------------\n');

        break;
      case 'CREATE_SUPPLY':
        console.log('-----------------------------------------------------------');

        conf.blockchainProperties.activeUser = {
          address: action.data.assetOwner, privateKey: action.data.assetOwnerPK,
        };

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly;
            break;
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly;
            break;
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily;
            break;
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly;
            break;
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro;
            break;
          case 'USD':
            currency = GeneralLib.Currency.USD;
            break;
          case 'Ether':
            currency = GeneralLib.Currency.Ether;
            break;
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar;
            break;
        }

        const supplyOffChainProperties: Market.Supply.SupplyOffchainProperties = {
          price: action.data.price,
          currency,
          availableWh: action.data.availableWh,
          timeframe: timeFrame,
        };

        const supplyProps: Market.Supply.SupplyOnChainProperties = {
          url: '',
          propertiesDocumentHash: '',
          assetId: action.data.assetId,
        };

        try {
          const supply = await Market.Supply.createSupply(supplyProps, supplyOffChainProperties, conf);
          delete supply.proofs;
          delete supply.configuration;
          conf.logger.info('Onboarded Supply ID: ' + supply.id);
        } catch (e) {
          conf.logger.error('Could not onboard a supply\n' + e);
        }

        console.log('-----------------------------------------------------------\n');

        break;

      case 'MAKE_AGREEMENT':
        console.log('-----------------------------------------------------------');

        conf.blockchainProperties.activeUser = {
          address: action.data.creator,
          privateKey: action.data.creatorPK,
        };

        switch (action.data.timeframe) {
          case 'yearly':
            timeFrame = GeneralLib.TimeFrame.yearly;
            break;
          case 'monthly':
            timeFrame = GeneralLib.TimeFrame.monthly;
            break;
          case 'daily':
            timeFrame = GeneralLib.TimeFrame.daily;
            break;
          case 'hourly':
            timeFrame = GeneralLib.TimeFrame.hourly;
            break;
        }

        switch (action.data.currency) {
          case 'Euro':
            currency = GeneralLib.Currency.Euro;
            break;
          case 'USD':
            currency = GeneralLib.Currency.USD;
            break;
          case 'Ether':
            currency = GeneralLib.Currency.Ether;
            break;
          case 'SingaporeDollar':
            currency = GeneralLib.Currency.SingaporeDollar;
            break;
        }

        const agreementOffchainProps: Market.Agreement.AgreementOffChainProperties = {
          start: action.data.startTime,
          ende: action.data.endTime,
          price: action.data.price,
          currency,
          period: action.data.period,
          timeframe: timeFrame,
        };

        const matcherOffchainProps: Market.Agreement.MatcherOffchainProperties = {
          currentWh: action.data.currentWh,
          currentPeriod: action.data.currentPeriod,
        };

        const agreementProps: Market.Agreement.AgreementOnChainProperties = {
          propertiesDocumentHash: null,
          url: null,
          matcherDBURL: null,
          matcherPropertiesDocumentHash: null,
          demandId: action.data.demandId,
          supplyId: action.data.supplyId,
          allowedMatcher: [action.data.allowedMatcher],
        };

        try {
          const agreement = await Market.Agreement.createAgreement(agreementProps, agreementOffchainProps, matcherOffchainProps, conf);
          delete agreement.proofs;
          delete agreement.configuration;
          delete agreement.propertiesDocumentHash;
          delete agreement.matcherPropertiesDocumentHash;
          if (agreement.approvedBySupplyOwner && agreement.approvedByDemandOwner) {
            conf.logger.info('Agreement Confirmed');
          }
          else if (!agreement.approvedByDemandOwner) {
            conf.logger.info('Demand Owner did not approve yet');
          }
          else if (!agreement.approvedBySupplyOwner) {
            conf.logger.info('Supply Owner did not approve yet');
          }
        } catch (e) {
          conf.logger.error('Error making an agreement\n' + e);
        }

        console.log('-----------------------------------------------------------\n');
        break;
      case 'APPROVE_AGREEMENT':
        console.log('-----------------------------------------------------------');

        conf.blockchainProperties.activeUser = {
          address: action.data.agree,
          privateKey: action.data.agreePK,
        };

        try {
          let agreement: Market.Agreement.Entity = await (new Market.Agreement.Entity(action.data.agreementId.toString(), conf)).sync();
          await agreement.approveAgreementSupply();
          agreement = await agreement.sync();
          if (agreement.approvedBySupplyOwner && agreement.approvedByDemandOwner) {
            conf.logger.info('Agreement Confirmed');
          }
          else if (!agreement.approvedByDemandOwner) {
            conf.logger.info('Demand Owner did not approve yet');
          }
          else if (!agreement.approvedBySupplyOwner) {
            conf.logger.info('Supply Owner did not approve yet');
          }
        } catch (e) {
          conf.logger.error('Could not approve agreement\n' + e);
        }

        console.log('-----------------------------------------------------------\n');
        break;
      default:
        const passString = JSON.stringify(action);
        await certificateDemo(passString, conf, adminPK);
    }
  }
  conf.logger.info('Total Time: ' + ((Date.now() - startTime) / 1000) + ' seconds');
};
